### PR TITLE
KOA-5828 Migrate colour token generation to compose

### PR DESCRIPTION
**Pull request by marianeum**
Moved the colour generation to compose (and deprecated colour warning) and cleaned up the existing tasks/dependencies

+ [x] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-android/blob/main/CONTRIBUTING.md)

Remember to include the following changes:
+ [x] Component `README.md`
+ [x] Tests
+ [x] Docs (either update [backpack-docs](https://github.com/Skyscanner/backpack-docs) now, or create a follow up ticket)

If you are curious about how we review, please read through the [code review guidelines](https://github.com/Skyscanner/backpack/blob/main/CODE_REVIEW_GUIDELINES.md)

### CODE REVIEW

File: .detekt.yml [hunk 1]
```diff
@@ -248,7 +248,6 @@ formatting:
   SpacingAroundComma:
     active: true
     autoCorrect: true
-    excludes: [ '**/tokens/**' ]
```
Comment (marianeum) [inline .detekt.yml:251]: fixed this one, though as it was actually our mistake in the generation

File: Backpack/src/main/res/values-night/backpack.semantic.color.xml [hunk 2]
```diff
@@ -18,152 +18,44 @@
...
-	<color name="__privateButtonPrimaryOnDarkPressedBackground">#ffc2c9cd</color>
-
-	<color name="__privateSkeletonShimmerStartEnd">#00000000</color>
-
-	<color name="__privateSkeletonShimmerCenter">#33000000</color>
-
- </resources>
+  <!-- THIS FILE IS AUTO GENERATED. DO NOT MODIFY! -->
+
+    <color name="bpkBackground">#ff010913</color>
```
Comment (marianeum) [inline Backpack/src/main/res/values-night/backpack.semantic.color.xml:23]: the only thing that changed is the order (and the private tokens extracted) - i actually double checked that (ie didn't sort), but as the compose version is sorted already this makes it easier to use

File: buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkColors.kt [hunk 3]
```diff
@@ -38,7 +38,14 @@ data class BpkColorModel(
   val darkReference: String?,
   val darkValue: String?,
   val deprecated: Boolean,
-)
+) {
+
+  fun value(isLight: Boolean) = if (isLight) {
```
Comment (marianeum) [inline buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkColors.kt:43]: moved this here so we can reuse it for compose + xml

File: buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkColors.kt [hunk 4]
```diff
@@ -48,7 +55,7 @@ object BpkColor {
 
     override fun invoke(source: Map<String, Any>): BpkColors =
       parseColors(source, resolveReferences = false) {
-        !it.isSemanticColor() && !it.isMarcomms()
+        !it.isSemanticColor() && !it.hasSemanticSuffix(newSemanticSuffixes) && !it.isMarcomms()
```
Comment (marianeum) [inline buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkColors.kt:58]: this is now only used in xml - we need to make sure we're filtering the right colours as it's slightly different to how the compose version was

File: buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkColors.kt [hunk 5]
```diff
@@ -58,16 +65,10 @@ object BpkColor {
       parseColors(source, resolveReferences = true) { it.isSemanticColor() && !it.isPrivate }.toBpkColors()
   }
 
-  object Internal : BpkParser<Map<String, Any>, Map<String, BpkColors>> {
+  object Internal : BpkParser<Map<String, Any>, BpkColors> {
```
Comment (marianeum) [inline buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkColors.kt:68]: moved the mapping to the formatter to keep the xml version simple/in one file

File: buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkColors.kt [hunk 6]
```diff
@@ -77,11 +78,25 @@ object BpkColor {
         toSemanticCompose(source, className)
     }
 
-    object InternalCompose : Format<Map<String, BpkColors>, List<TypeSpec>>() {
-      override fun invoke(source: Map<String, BpkColors>): List<TypeSpec> =
-        source.map { toInternalCompose(it.value, it.key) }
+    object InternalCompose : Format<BpkColors, List<TypeSpec>>() {
```
Comment (marianeum) [inline buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkColors.kt:81]: that's just all stuff that moved from the parser

File: buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkColors.kt [hunk 7]
```diff
@@ -170,8 +186,10 @@ private fun deprecationProperty(): PropertySpec =
     .addModifiers(KModifier.CONST, KModifier.PRIVATE)
     .build()
 
+private fun String.toArgb() = substring(7) + substring(1, 7)
```
Comment (marianeum) [inline buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkColors.kt:189]: extracted to reuse in xml

File: buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkOutput.kt [hunk 8]
```diff
@@ -78,18 +78,32 @@ sealed class BpkOutput<Input> : (Input) -> Boolean {
     val name: String,
   ) : BpkOutput<String>() {
     override fun invoke(content: String): Boolean {
-      val target = File(File(srcDir, folder), "backpack.$name.xml")
-      target.createNewFile()
-
-      val template = Resources.toString(Resources.getResource("resource_file_template.txt"), StandardCharsets.UTF_8)
-
-      target.writeText(template.replace("{{content}}", content))
+      content.writeToFile(srcDir, folder, name)
```
Comment (marianeum) [inline buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkOutput.kt:81]: extracted for reuse

File: buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkOutput.kt [hunk 9]
```diff
@@ -78,18 +78,32 @@ sealed class BpkOutput<Input> : (Input) -> Boolean {
...
       return true
     }
+  }
 
+  data class XmlFiles(
```
Comment (marianeum) [inline buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkOutput.kt:86]: this is to support the day + night mode variation

File: buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkColors.kt (hunk 5 above)
Comment (bvitaliyg) [reply buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkColors.kt:68]: We may be needed to re-package it in the future

File: gulpfile.js [hunk 10]
```diff
@@ -91,110 +70,6 @@ const convertToXml = (chunk, enc, cb) => {
...
-  const getColors = () =>
-    tokensWithType('color')
-      .map(color => {
-        const colorObject = JSON.parse(JSON.stringify(color));
-        colorObject.name = `bpk${pascalCase(colorObject.name.replace(colorObject.type.toUpperCase(), ''))}`;
-        return colorObject;
-      })
-      .filter(entry => entry.deprecated);
-
-  return gulp
```
Comment (bvitaliyg) [inline gulpfile.js:187]: Veery happy to see this

File: package.json [hunk 11]
```diff
@@ -42,20 +42,16 @@
...
     "danger": "^11.0.2",
-    "del": "^7.0.0",
     "eslint": "^8.11.0",
     "eslint-config-airbnb": "^19.0.0",
     "eslint-config-prettier": "^8.3.0",
     "eslint-plugin-eslint-comments": "^3.2.0",
     "eslint-plugin-prettier": "^4.0.0",
     "gulp": "^4.0.2",
-    "gulp-nunjucks": "^5.1.0",
-    "gulp-rename": "^2.0.0",
```
Comment (bvitaliyg) [inline package.json:53]: And this

**Review by bvitaliyg — APPROVED** [review]

File: buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkColors.kt (hunk 5 above)
Comment (marianeum) [reply buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkColors.kt:68]: yep easy enough to do if we need to - it's better now than it was before - at least now the private ones are in a separate file

File: buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkColors.kt [hunk 12]
```diff
@@ -303,3 +315,29 @@ private fun toInternalCompose(
...
+    CaseFormat.UPPER_UNDERSCORE.to(CaseFormat.UPPER_CAMEL, this)
+
+  fun BpkColorModel.name() =
+    if (isPrivate) {
+      "__private${name.toCamelCase()}"
+    } else {
+      "bpk${name.toCamelCase()}"
+    }
+
+  return source.joinToString("\n") { model ->
```
[the remark this answers is not in the document]
Comment (marianeum) [reply buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkColors.kt:330]: yep. same as before - the resources files are very simple, so an xml generator would be overkill

**Pull request merged**